src/ucp/core/ucp_mm.c

```diff
@@ -460,13 +427,65 @@ static ucs_status_t ucp_mem_map_common(ucp_context_h context, void *address,
                                        unsigned uct_flags, int is_allocate,
                                        const char *alloc_name, ucp_mem_h *memh_p)
 {
-    return ucp_memh_get_common(context, address, length, uct_flags, memory_type,
-                               context->reg_md_map[memory_type], memh_p,
-                               is_allocate, alloc_name);
-}
+    ucp_md_map_t reg_md_map       = context->reg_md_map[memory_type];
+    ucp_md_index_t alloc_md_index = UCP_NULL_RESOURCE;
+    ucp_mem_h memh                = NULL;
+    ucp_md_index_t md_index;
+    uct_allocated_memory_t mem;
+    ucs_status_t status;
+    ucs_memory_type_t reg_mem_type;
+    void *reg_address;
+    size_t reg_length;
+
+    if (is_allocate) {
+        status = ucp_mem_alloc(context, address, length, uct_flags, memory_type,
+                               alloc_name, &mem);
+        if (status != UCS_OK) {
+            return status;
+        }
 
-int ucp_memh_zero_length(const ucp_mem_h memh) {
-    return memh == &ucp_mem_dummy_handle;
+        reg_address  = mem.address;
+        reg_length   = mem.length;
+        reg_mem_type = mem.mem_type;
+
+        if (mem.method == UCT_ALLOC_METHOD_MD) {
+            for (md_index = 0; md_index < context->num_mds; md_index++) {
+                if (mem.md == context->tl_mds[md_index].md) {
+                    alloc_md_index = md_index;
+                    reg_md_map    &= ~UCS_BIT(md_index);
+                    break;
+                }
+            }
+
+            ucs_assert(alloc_md_index != UCP_NULL_RESOURCE);
+        }
+    } else {
+        reg_address  = address;
+        reg_length   = length;
+        reg_mem_type = memory_type;
+    }
+
+    status = ucp_memh_get_slow(context, reg_address, reg_length,
+                               reg_mem_type, reg_md_map, uct_flags, &memh);
+    if (status != UCS_OK) {
+        return status;
+    }
```

Comment:
it looks like this is the only part common for allocate and register cases, and also register case is only from one place in UCP - ucp_mem_map API call.
So maybe transform ucp_mem_map_common() to ucp_mem_allocate(), and the one place that wants only to register -- will call ucp_memh_get() directly ?